File: HW4-PCA/matlab/Hw4CompEx.m
% EECS2050 Linear Algebra
% Computer assignment for HW#4
% Principal value decomposition
% Due Aug 15, 2022
% Prof. Yi-Wen Liu


clear; close all
N = 50; % number of datapoints in 2D
rng(1);% random seed

% randomly create two correlated columns
A = randn(N,2);
B = [0.5 0.5; 0.6 -0.1];
X = A*B;

figure(1);
plot(X(:,1), X(:,2),'x');
grid on; pbaspect([1 1 1])

% Next, let us find a change of coordinates to diagonalize the covariance matrix.
mu = mean(X);
X_new = X - repmat(mu,N,1);

% >>>> TODO: Use diagonalization to get the pinciple axes.
% Q is the eigenvector"s"
% eg. Q(:,1) is the first eigenvector 
% D is diagonal matrix with eigenvalues

Sigma = 1/N* X_new' * X_new; % expected value of the covariance matrix
[Q,D] = eig(Sigma); % <----- This line is the key step for diagonalization.
prinAxes = Q*sqrt(D);

% <<< End of TODO

%%  Decorrelation by projecting onto principal components, and add the mean back.
Y = X_new*Q + repmat(mu,N,1);
figure(1); hold on;
plot(Y(:,1), Y(:,2),'.');

% Draw the principle axes
h1 = line(mu(1)+[0 prinAxes(1,1)],mu(2)+[0 prinAxes(2,1)],'color','k');
h2 = line(mu(1)+[0 prinAxes(1,2)],mu(2)+[0 prinAxes(2,2)],'color','k');
set(h1,'linewidth',2);
set(h2,'linewidth',2);

% set to aspect ratio to 1:1
set(gcf,'position',[400 300 500 500]); 
set(gca,'position',[0.1 0.1 0.8 0.8]);
set(gca,'xlim',[-3 3]);
set(gca,'ylim',[-3 3]);


%% Total Error with Principle components
% When we project the points on different axes, 
% we will get different errors.
% This part will show you the total errors 
% when we project the points on the two principle components.

m_axe1 = prinAxes(2,1) / prinAxes(1,1);
m_axe2 = prinAxes(2,2) / prinAxes(1,2);

figure(2);
% plot Errors with Axe1
subplot(121);
scatter(X(:,1), X(:,2),'x');
hold on; grid on; pbaspect([1 1 1]); 

line([-3 3],[-3*m_axe1 3*m_axe1],'color','#A2142F');
dis_sum = 0;
for kk = 1:N
    P = X(kk, 1:2);
    x0 = P(1); y0 = P(2);
    
    [dis, x1, y1] = get_error(x0, y0, m_axe1);
    scatter(x0,y0,"green");
    scatter(x1,y1,"green");
    dis_sum = dis_sum + dis;
    h1 = line([x0, x1], [y0, y1], "color", "#4DBEEE");
    set(h1,'linewidth',0.1);
%     pause;
end
title(["Total Error with PC_1", num2str(dis_sum)]);
xlim([-2, 2]);
ylim([-2, 2]);
pbaspect([1 1 1]);

% plot Errors with Axe2
subplot(122);
scatter(X(:,1), X(:,2),'x');
hold on; grid on; 

line([-3 3],[-3*m_axe2 3*m_axe2],'color','#A2142F');

% calculate the total errors
dis_sum = 0;
for kk = 1:N
    P = X(kk, 1:2);
    x0 = P(1); y0 = P(2);
    
    [dis, x1, y1] = get_error(x0, y0, m_axe2);
    scatter(x0,y0,"green");
    scatter(x1,y1,"green");
    dis_sum = dis_sum + dis;
    h1 = line([x0, x1], [y0, y1], "color", "#4DBEEE");
    set(h1,'linewidth',0.1);
%     pause;
end
title(["Total Error with PC_2", num2str(dis_sum)]);
xlim([-2, 2]);
ylim([-2, 2]);
pbaspect([1 1 1]);

%% Total Error with arbitary axes
% This part will show you how the total Error change 
% when we change the axes.
% NOTE: If you feel the process is too slow, 
%       you can reduce the size N or num_degrees.

num_degrees = 100; % how many graphs we calculate

thetas = linspace(0,2*pi, num_degrees);
figure(3);
for jj=1:length(thetas)
    hold off;
    scatter(X(:,1), X(:,2),'x');
    hold on; grid on;

    % Rotate the axes
    theta = thetas(jj);
    [a_x, a_y] = pol2cart(theta, 3);
    [b_x, b_y] = pol2cart(theta + pi, 3);
    m_k = a_y/ a_x;
    line([a_x, b_x], [a_y b_y],'color','#A2142F');
    
    % calculate the total errors
    dis_sum = 0;
    for kk = 1:N
        P = X(kk, 1:2);
        x0 = P(1); y0 = P(2);
        
        [dis, x1, y1] = get_error(x0, y0, m_k);
        scatter(x0,y0,"green");
        scatter(x1,y1,"green");
        dis_sum = dis_sum + dis;
        h1 = line([x0, x1], [y0, y1], "color", "#4DBEEE");
        set(h1,'linewidth',0.1);
    end
    title(["Total Error", num2str(dis_sum)]);
    
    xlim([-2, 2]);
    ylim([-2 2]);
%     pause;
    drawnow;
end

File: HW4-PCA/python/HW4CompEx.py
#%%
import numpy as np
import matplotlib.pyplot as plt
from matplotlib import collections
import os
#%% Random Generate data
N = 250
np.random.seed(1)
A = np.random.normal(size=[N,2])
B = np.array([[0.5, 0.5], [0.6, -0.1]])
X = np.matmul(A,B) # randomly create two correlated columns
#%% find a change of coordinates to diagonalize the covariance matrix.
mu = np.mean(X, axis=0)
X_new = X - np.tile(mu, [N,1])

#%% Finish this part 
# TODO: Use diagonalization to get the pinciple axes.
# Sigma = ... <-expected value of the covariance matrix
# Q = ... <-eigenvectors matrix, Q[:,0] is the first eigenvectors
# D = ... <-diagonal matrix with eigenvalues
# prinAxes = ... <-Reconstruct the pinciple axes
Sigma = 1/N * np.matmul(X_new.T, X_new); 
w, Q = np.linalg.eig(Sigma)
D = np.array([[w[0], 0], [0, w[1]]])
prinAxes = np.matmul(Q, np.sqrt(D))
#%% 
Y = np.matmul(X_new, Q) + np.tile(mu, [N,1])
plt.scatter(X[:,0], X[:,1], s=3, c="blue", marker="8")
plt.scatter(Y[:,0], Y[:,1], s=3, c="green", marker="v")
plt.plot([0,prinAxes[0,0]], [0, prinAxes[1,0]], linewidth=2, color="red")
plt.plot([0,prinAxes[0,1]], [0, prinAxes[1,1]], linewidth=2, color="red")
plt.xlim([-3, 3])
plt.ylim([-3, 3])
plt.show()

#%% Total Error with Principle components
# When we project the points on different axes, 
# we will get different errors.
# This part will show you the total errors 
# when we project the points on the two principle components.

def get_error(x0, y0,m):
    """ Get Distance and Location with Line y=ax and Point (x0,y0)

        dis: The minimal distance bewteen (x0,y0) and y=ax.
        (x1, y1): The point in y=ax with minimal distance.
    """
    x1 = (y0*m + x0)/(m*m+1)
    y1 = x1*m
    dis = (x0 - x1)**2 + (y0 - y1)**2
    return dis, x1, y1 
``
m_axe1 = prinAxes[1,0] / prinAxes[0,0]
m_axe2 = prinAxes[1,1] / prinAxes[0,1]

# Get each distance of each point
x0s = X[:,0]
y0s = X[:,1]
# PC_1
ans = np.apply_along_axis(lambda P: get_error(x0=P[0], y0=P[1], m=m_axe1), axis=1, arr=X)
x1s,y1s,dis1 = ans[:,1], ans[:,2], ans[:,0]
lines1 = [[(x0, y0), (x1, y1)] for x0, y0, x1, y1 in zip(x0s, y0s, x1s, y1s)]
# PC_2
ans = np.apply_along_axis(lambda P: get_error(x0=P[0], y0=P[1], m=m_axe2), axis=1, arr=X)
x1s,y1s,dis2 = ans[:,1], ans[:,2], ans[:,0]
lines2 = [[(x0, y0), (x1, y1)] for x0, y0, x1, y1 in zip(x0s, y0s, x1s, y1s)]

# %%
fig,axes=plt.subplots(1,2)
for index, (lines, m, dis) in enumerate([(lines1, m_axe1, dis1), (lines2, m_axe2, dis2)]):
    segments = collections.LineCollection(lines, color="#2A9D8F", linewidth = 0.5)
    axes[index].add_collection(segments)
    axes[index].plot([-3, 3], [-3*m, 3*m], c="#E9C46A")
    axes[index].scatter(x0s, y0s, s=3, c="#E76F51")
    axes[index].set_aspect(1)
    axes[index].set_xlim([-2,2])
    axes[index].set_ylim([-2,2])
    axes[index].set_title(f"Total Error with PC_{index+1}: \n{sum(dis):.2f}")
plt.show()

# %% Total Error with arbitary axes
# This part will show you how the total Error change 
# when we change the axes.
# NOTE: If you feel the process is too slow, 
#       you can reduce the size N or num_degrees.
from pathlib import Path
Path("output").mkdir(exist_ok=True)
num_degrees = 180
thetas = np.deg2rad(np.linspace(0,360,num_degrees))
ms = np.tan(thetas) # m = y/x = tan(theta)
for index, m in enumerate(ms):
    fig, axes = plt.subplots()
    ans = np.apply_along_axis(lambda P: get_error(x0=P[0], y0=P[1], m=m), axis=1, arr=X)
    x1s,y1s,dis = ans[:,1], ans[:,2], ans[:,0]
    lines = [[(x0, y0), (x1, y1)] for x0, y0, x1, y1 in zip(x0s, y0s, x1s, y1s)]
    segments = collections.LineCollection(lines, color="#2A9D8F", linewidth = 0.5)
    axes.add_collection(segments)
    axes.plot([-3, 3], [-3*m, 3*m], c="#E9C46A")
    axes.scatter(x0s, y0s, s=3, c="#E76F51")
    axes.set_aspect(1)
    axes.set_xlim([-2,2])
    axes.set_ylim([-2,2])
    axes.set_title(f"Total Error:\n{sum(dis):.2f}")
    plt.savefig(f"output/{index}.png")
    plt.show()
# %%
# plot with gif animation
import imageio
from glob import glob
with imageio.get_writer("Error change.gif", mode="I") as writer:
    for i in range(num_degrees):
        filename = f"output/{i}.png"
        img = imageio.imread(filename)
        writer.append_data(img)

# %%
